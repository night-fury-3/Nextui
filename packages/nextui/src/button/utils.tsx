import React from 'react';
import { ButtonProps } from '../button/button';
import { ButtonGroupConfig } from '../button/button-group-context';

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    ghost: config.ghost || props.ghost,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
