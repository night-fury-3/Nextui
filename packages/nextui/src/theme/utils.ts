import { NextUIThemes } from './index';
import useWarning from '@hooks/use-warning';
import { DeepPartial } from '@utils/types';
import { deepMergeObject } from '@utils/object';
import darkTheme from './dark';
import lightTheme from './default';

export type PartialTheme = DeepPartial<NextUIThemes>;

export type ThemeParam =
  | PartialTheme
  | ((theme: PartialTheme) => PartialTheme)
  | undefined;

export const mergeTheme = (
  current: NextUIThemes,
  custom: ThemeParam
): NextUIThemes => {
  if (!custom) return current;
  if (typeof custom === 'function') {
    const merged = custom(current);
    if (!merged || typeof merged !== 'object') {
      useWarning('The theme function must return object value.');
    }
    return merged as NextUIThemes;
  }
  return deepMergeObject<NextUIThemes>(current, custom as NextUIThemes);
};

export const switchTheme = (mergedTheme: NextUIThemes): NextUIThemes => {
  const themes: { [key in NextUIThemes['type']]: NextUIThemes } = {
    light: lightTheme,
    dark: darkTheme,
  };
  return { ...mergedTheme, ...themes[mergedTheme.type] };
};
