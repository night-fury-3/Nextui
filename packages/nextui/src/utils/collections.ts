import React, { ReactNode } from 'react';

export const getId = () => {
  return Math.random().toString(32).slice(2, 10);
};

export const hasChild = (
  children: ReactNode | undefined,
  child: React.ElementType
): boolean => {
  const types = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return null;
    return item.type;
  });

  return (types || []).includes(child);
};

export const pick = <Obj extends { [key: string]: any }, Key extends keyof Obj>(
  props: Key[],
  obj: Obj
): Pick<Obj, Key> =>
  props.reduce((acc, prop) => {
    acc[prop] = obj[prop];
    return acc;
  }, {} as Pick<Obj, Key>);

export const pickChild = (
  children: ReactNode | undefined,
  targetChild: React.ElementType
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = [];
  const withoutTargetChildren = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return item;
    if (item.type === targetChild) {
      target.push(item);
      return null;
    }
    return item;
  });

  const targetChildren = target.length >= 0 ? target : undefined;

  return [withoutTargetChildren, targetChildren];
};
