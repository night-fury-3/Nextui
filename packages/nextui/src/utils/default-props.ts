import { NormalSizes } from './prop-types';
import type { CSSProperties } from 'react';

export type NextUISpacingWithAuto = NormalSizes | 'px' | 'auto' | number;

export interface NextUIMargins {
  m?: NextUISpacingWithAuto;
  my?: NextUISpacingWithAuto;
  mx?: NextUISpacingWithAuto;
  mt?: NextUISpacingWithAuto;
  mb?: NextUISpacingWithAuto;
  ml?: NextUISpacingWithAuto;
  mr?: NextUISpacingWithAuto;
}

export interface NextUIPaddings {
  p?: NextUISpacingWithAuto;
  py?: NextUISpacingWithAuto;
  px?: NextUISpacingWithAuto;
  pt?: NextUISpacingWithAuto;
  pb?: NextUISpacingWithAuto;
  pl?: NextUISpacingWithAuto;
  pr?: NextUISpacingWithAuto;
}

export type NextUISpaces = NextUIMargins & NextUIPaddings;

export type NextUISpacesKeys = { [key in keyof NextUISpaces]?: string };

export const NextUISpacesStringKeys = [
  'm',
  'my',
  'mx',
  'mt',
  'mb',
  'ml',
  'mr',
  'p',
  'py',
  'px',
  'pt',
  'pb',
  'pl',
  'pr'
];

export interface DefaultProps extends NextUIMargins, NextUIPaddings {
  style?: CSSProperties;
  className?: string;
}
