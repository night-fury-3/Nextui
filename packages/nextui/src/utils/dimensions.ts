import { NormalWeights, NormalSizes } from './prop-types';

interface GetSizeValue {
  size: string | number;
  sizes: Record<string, any>;
  defaultSize?: string;
}

export const getMargin = (num: number): string => {
  return `calc(${num * 15.25}pt + 1px * ${num - 1})`;
};

export const getNormalWeight = (weight?: NormalWeights): string | undefined => {
  const weights: { [key in NormalWeights]?: string } = {
    light: '1px',
    normal: '2px',
    bold: '3px'
  };
  return weights[weight || 'normal'];
};

export const getNormalRadius = (
  size: NormalSizes,
  rounded?: boolean
): string => {
  const radius: { [key in NormalSizes]: string } = {
    mini: '7px',
    small: '9px',
    medium: '12px',
    large: '14px',
    xlarge: '15px'
  };
  const baseRadius = radius[size];
  return rounded ? `calc(${baseRadius} + 10rem)` : baseRadius;
};

export function getSizeValue({
  size,
  sizes,
  defaultSize = 'md'
}: GetSizeValue) {
  if (typeof size === 'number') {
    return size;
  }
  return sizes[size] || size || sizes[defaultSize];
}
