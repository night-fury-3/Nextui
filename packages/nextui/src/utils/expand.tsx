import React, { useEffect, useRef, useState } from 'react';
import withDefaults from '../utils/with-defaults';
import useRealShape from '../use-real-shape';
import clsx from './clsx';

interface Props {
  isExpanded?: boolean;
  animated?: boolean;
  delay?: number;
}

const defaultProps = {
  isExpanded: false,
  animated: true,
  delay: 200
};

export type ExpandProps = Props & typeof defaultProps;

const Expand: React.FC<React.PropsWithChildren<ExpandProps>> = ({
  isExpanded,
  delay,
  animated,
  children
}) => {
  const [height, setHeight] = useState<string>(isExpanded ? 'auto' : '0');
  const [selfExpanded, setSelfExpanded] = useState<boolean>(isExpanded);
  const contentRef = useRef<HTMLDivElement>(null);
  const entryTimer = useRef<number>();
  const leaveTimer = useRef<number>();
  const resetTimer = useRef<number>();
  const [state, updateShape] = useRealShape<HTMLDivElement>(contentRef);

  useEffect(() => setHeight(`${state.height}px`), [state.height]);
  useEffect(() => {
    // show element or reset height.
    // force an update once manually, even if the element does not change.
    // (the height of the element might be "auto")
    if (!isExpanded) {
      updateShape();
      setHeight(`${state.height}px`);
    }

    // show expand animation
    entryTimer.current = window.setTimeout(() => {
      setSelfExpanded(isExpanded);
      clearTimeout(entryTimer.current);
    }, 30);

    // Reset height after animation
    if (isExpanded) {
      resetTimer.current = window.setTimeout(() => {
        setHeight('auto');
        clearTimeout(resetTimer.current);
      }, delay);
    } else {
      leaveTimer.current = window.setTimeout(() => {
        clearTimeout(leaveTimer.current);
      }, delay / 2);
    }

    return () => {
      clearTimeout(entryTimer.current);
      clearTimeout(leaveTimer.current);
      clearTimeout(resetTimer.current);
    };
  }, [isExpanded]);

  return (
    <div className={clsx('container', { expanded: selfExpanded, animated })}>
      <div ref={contentRef} className="content">
        {children}
      </div>
      <style jsx>{`
        .container {
          padding: 0;
          margin: 0;
          height: 0;
          opacity: 0;
          overflow: hidden;
        }
        .animated {
          transition: height ${delay}ms ease 0ms,
            opacity ${delay * 1.5}ms ease 0ms;
        }
        .expanded {
          height: ${height};
          opacity: 1;
        }
      `}</style>
    </div>
  );
};

export default withDefaults(Expand, defaultProps);
