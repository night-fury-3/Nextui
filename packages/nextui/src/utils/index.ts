import { NormalColors, normalColors } from './prop-types';
import { NextUIThemesPalette } from '@theme/index';
/**
 * This function allows validate if a string is a hexadecimal
 * value
 * @param str [string] hexadecimal value
 * @returns result [boolean]
 */
export const isHex = (str: string): boolean => {
  const exp = /#[a-fA-F0-9]{3,6}/g;
  return exp.test(str);
};

/**
 *
 * @param hex [string]
 * @param alpha [number]
 * @returns [string]
 */
export const hexToRGBA = (hex: string, alpha: number = 1): string => {
  let r: string | number = 0,
    g: string | number = 0,
    b: string | number = 0;
  // 3 digits
  if (hex.length == 4) {
    r = '0x' + hex[1] + hex[1];
    g = '0x' + hex[2] + hex[2];
    b = '0x' + hex[3] + hex[3];
    // 6 digits
  } else if (hex.length == 7) {
    r = '0x' + hex[1] + hex[2];
    g = '0x' + hex[3] + hex[4];
    b = '0x' + hex[5] + hex[6];
  }
  return `rgba(${+r}, ${+g},${+b},${alpha})`;
};

export const getNormalColor = (
  color: NormalColors | string | undefined,
  palette: NextUIThemesPalette,
  defaultColor: string = 'inherit'
) => {
  const colors: { [key in NormalColors | string]: string } = {
    default: defaultColor,
    primary: palette.primary,
    secondary: palette.secondary,
    success: palette.success,
    warning: palette.warning,
    error: palette.error,
    gradient: palette.gradient,
  };
  return color && colors[color] ? colors[color] : defaultColor;
};

export const isNormalColor = (color: string): boolean => {
  let found = normalColors.find((el) => el === color);
  return found !== undefined && found !== null;
};

/**
 * Function that checks color name support in the current browser
 * @param strColor
 * @returns boolean
 */
export const isColor = (strColor: string) => {
  let s = new Option().style;
  s.color = strColor;
  return s.color == strColor;
};
