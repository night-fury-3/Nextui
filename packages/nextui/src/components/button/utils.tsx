import React, { ReactNode } from 'react';
import { NormalSizes } from '@utils/prop-types';
import ButtonIcon from './button-icon';
import { ButtonProps } from '@components/button/button';
import { ButtonGroupConfig } from '@components/button/button-group-context';

export const getButtonChildrenWithIcon = (
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode;
    iconRight?: React.ReactNode;
  }
) => {
  const { icon, iconRight } = icons;
  const hasIcon = icon || iconRight;
  const isRight = Boolean(iconRight);
  const paddingForAutoMode =
    auto || size === 'mini'
      ? `calc(var(--next-ui-button-height) / 2 + var(--next-ui-button-padding) * .5)`
      : 0;
  if (!hasIcon) return <div className="text">{children}</div>;
  if (React.Children.count(children) === 0) {
    return (
      <ButtonIcon isRight={isRight} isSingle>
        {hasIcon}
      </ButtonIcon>
    );
  }
  return (
    <React.Fragment>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </React.Fragment>
  );
};

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
