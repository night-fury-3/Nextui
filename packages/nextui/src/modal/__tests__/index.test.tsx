import React from 'react';
import { mount } from 'enzyme';
import Modal from '../index';
import { nativeEvent, updateWrapper } from 'tests/utils';
import { expectModalIsClosed, expectModalIsOpened } from './use-modal.test';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

const TabEvent = {
  key: 'TAB',
  keyCode: 9,
  which: 9
};

describe('Modal', () => {
  it('should render correctly', () => {
    const wrapper = mount(
      <Modal open={true}>
        <Modal.Header>Modal</Modal.Header>
        <Modal.Body>
          <p>Some content contained within the modal.</p>
        </Modal.Body>
        <Modal.Footer>Cancel</Modal.Footer>
      </Modal>
    );
    expect(() => wrapper.unmount()).not.toThrow();
  });
  it('should trigger event when modal changed', async () => {
    const openHandler = jest.fn();
    const closeHandler = jest.fn();
    const wrapper = mount(
      <Modal onOpen={openHandler} onClose={closeHandler}>
        <Modal.Header>Modal</Modal.Header>
      </Modal>
    );
    expectModalIsClosed(wrapper);
    wrapper.setProps({ open: true });
    await updateWrapper(wrapper, 350);
    expectModalIsOpened(wrapper);
    expect(openHandler).toHaveBeenCalled();
    wrapper.find('.nextui-backdrop').simulate('click', nativeEvent);
    await updateWrapper(wrapper, 500);
    expectModalIsClosed(wrapper);
    expect(closeHandler).toHaveBeenCalled();
  });
  it('should disable backdrop event', async () => {
    const closeHandler = jest.fn();
    const wrapper = mount(
      <Modal open={true} preventClose onClose={closeHandler}>
        <Modal.Header>Modal</Modal.Header>
        <Modal.Footer>Submit</Modal.Footer>
      </Modal>
    );
    wrapper.find('.nextui-backdrop').simulate('click', nativeEvent);
    await updateWrapper(wrapper, 500);
    expectModalIsOpened(wrapper);
    expect(closeHandler).not.toHaveBeenCalled();
  });
  it('customization should be supported', () => {
    const wrapper = mount(
      <Modal open={true} width="100px" className="test-class">
        <Modal.Header>Modal</Modal.Header>
      </Modal>
    );
    const html = wrapper.find('.nextui-modal-wrapper').html();
    expect(html).toContain('test-class');
    expect(() => wrapper.unmount()).not.toThrow();
  });
  it('focus should only be switched within modal', () => {
    const wrapper = mount(
      <Modal open={true} width="100px" className="test-class">
        <Modal.Header>Modal</Modal.Header>
      </Modal>
    );
    const tabStart = wrapper.find('.hide-tab').at(0).getDOMNode();
    const tabEnd = wrapper.find('.hide-tab').at(1).getDOMNode();
    const eventElement = wrapper.find('.nextui-modal-wrapper').at(0);
    expect(document.activeElement).toBe(tabStart);
    act(() => {
      eventElement.simulate('keydown', {
        ...TabEvent,
        shiftKey: true
      });
    });
    expect(document.activeElement).toBe(tabEnd);
    act(() => {
      eventElement.simulate('keydown', {
        ...TabEvent,
        shiftKey: false
      });
    });
    expect(document.activeElement).toBe(tabStart);
  });
  it('should close modal when keyboard event is triggered', async () => {
    const wrapper = mount(
      <Modal open={true}>
        <Modal.Header>Modal</Modal.Header>
      </Modal>
    );
    expectModalIsOpened(wrapper);
    userEvent.keyboard('{esc}');
    await updateWrapper(wrapper, 500);
    expectModalIsClosed(wrapper);
  });
});
