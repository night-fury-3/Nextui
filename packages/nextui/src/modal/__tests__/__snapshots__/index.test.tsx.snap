// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal customization should be supported 1`] = `
"<div class=\\"wrapper test-class  wrapper-enter\\" role=\\"dialog\\" aria-modal=\\"true\\" tabindex=\\"-1\\"><div tabindex=\\"0\\" class=\\"hide-tab\\" aria-hidden=\\"true\\"></div><div class=\\"modal-header auto-margin\\">Modal<style>
        .modal-header {
          display: flex;
          flex-shrink: 0;
          justify-content: center;
          align-items: center;
          overflow: hidden;
          color: inherit;
          background-color: inherit;
          font-size: 0.875rem;
          padding: 8pt calc(16pt + 0.25rem);
          border-bottom-left-radius: 14px;
          border-bottom-right-radius: 14px;
        }
        .auto-margin > :global(*:first-child) {
          margin-top: 0;
        }
        .auto-margin > :global(*:last-child) {
          margin-bottom: 0;
        }
      </style></div><div tabindex=\\"0\\" class=\\"hide-tab\\" aria-hidden=\\"true\\"></div><style>
          .wrapper {
            max-width: 100%;
            vertical-align: middle;
            overflow: hidden;
            display: flex;
            flex-direction: column;
            position: relative;
            box-sizing: border-box;
            background-color: #fff;
            color: #000;
            border-radius: 14px;
            box-shadow: 0 30px 60px rgba(0, 0, 0, 0.15);
            opacity: 0;
            outline: none;
            transform: translate3d(0px, 15px, 0px) scale(0.9);
            transition: opacity 0.25s cubic-bezier(0.4, 0, 0.2, 1) 0s,
              transform 0.25s cubic-bezier(0.4, 0, 0.2, 1) 0s;
          }
          .wrapper-enter {
            opacity: 0;
            transform: translate3d(0px, 15px, 0px) scale(0.9);
          }
          .wrapper-enter-active {
            opacity: 1;
            transform: translate3d(0px, 0px, 0px) scale(1);
          }
          .wrapper-leave {
            opacity: 1;
            transform: translate3d(0px, 0px, 0px) scale(1);
          }
          .wrapper-leave-active {
            opacity: 0;
            transform: translate3d(0px, 15px, 0px) scale(0.9);
          }
          .hide-tab {
            outline: none;
            overflow: hidden;
            width: 0;
            height: 0;
            opacity: 0;
          }
        </style></div>"
`;

exports[`Modal should render correctly 1`] = `
"<div class=\\"backdrop  backdrop-wrapper-enter\\"><div class=\\"layer\\"></div><div class=\\"content\\"><div class=\\"wrapper   wrapper-enter\\" role=\\"dialog\\" aria-modal=\\"true\\" tabindex=\\"-1\\"><div tabindex=\\"0\\" class=\\"hide-tab\\" aria-hidden=\\"true\\"></div><div class=\\"modal-header auto-margin\\">Modal<style>
        .modal-header {
          display: flex;
          flex-shrink: 0;
          justify-content: center;
          align-items: center;
          overflow: hidden;
          color: inherit;
          background-color: inherit;
          font-size: 0.875rem;
          padding: 8pt calc(16pt + 0.25rem);
          border-bottom-left-radius: 14px;
          border-bottom-right-radius: 14px;
        }
        .auto-margin > :global(*:first-child) {
          margin-top: 0;
        }
        .auto-margin > :global(*:last-child) {
          margin-bottom: 0;
        }
      </style></div><div class=\\"modal-body auto-margin\\"><p>Some content contained within the modal.</p></div><style>
        .modal-body {
          flex: 1 1 auto;
          padding: 8pt calc(16pt + 0.25rem);
          position: relative;
          text-align: left;
        }
        .auto-margin > :global(*:first-child) {
          margin-top: 0;
        }
        .auto-margin > :global(*) {
          margin-bottom: 1rem;
        }
        .auto-margin > :global(*:last-child) {
          margin-bottom: 0;
        }
      </style><div class=\\"modal-footer auto-margin\\">Cancel<style>
        .modal-footer {
          display: flex;
          flex-wrap: wrap;
          flex-shrink: 0;
          overflow: hidden;
          color: inherit;
          align-items: center;
          background-color: inherit;
          font-size: 0.875rem;
          padding: 8pt 16pt;
          justify-content: flex-end;
          border-bottom-left-radius: 14px;
          border-bottom-right-radius: 14px;
        }
        .auto-margin > :global(*) {
          margin: 0.25rem;
        }
      </style></div><div tabindex=\\"0\\" class=\\"hide-tab\\" aria-hidden=\\"true\\"></div><style>
          .wrapper {
            max-width: 100%;
            vertical-align: middle;
            overflow: hidden;
            display: flex;
            flex-direction: column;
            position: relative;
            box-sizing: border-box;
            background-color: #fff;
            color: #000;
            border-radius: 14px;
            box-shadow: 0 30px 60px rgba(0, 0, 0, 0.15);
            opacity: 0;
            outline: none;
            transform: translate3d(0px, 15px, 0px) scale(0.9);
            transition: opacity 0.25s cubic-bezier(0.4, 0, 0.2, 1) 0s,
              transform 0.25s cubic-bezier(0.4, 0, 0.2, 1) 0s;
          }
          .wrapper-enter {
            opacity: 0;
            transform: translate3d(0px, 15px, 0px) scale(0.9);
          }
          .wrapper-enter-active {
            opacity: 1;
            transform: translate3d(0px, 0px, 0px) scale(1);
          }
          .wrapper-leave {
            opacity: 1;
            transform: translate3d(0px, 0px, 0px) scale(1);
          }
          .wrapper-leave-active {
            opacity: 0;
            transform: translate3d(0px, 15px, 0px) scale(0.9);
          }
          .hide-tab {
            outline: none;
            overflow: hidden;
            width: 0;
            height: 0;
            opacity: 0;
          }
        </style></div></div><style>
            .backdrop {
              position: fixed;
              top: 0;
              left: 0;
              right: 0;
              bottom: 0;
              overflow: auto;
              z-index: 1000;
              -webkit-overflow-scrolling: touch;
              box-sizing: border-box;
              text-align: center;
            }

            .content {
              position: relative;
              z-index: 1001;
              outline: none;
              max-width: 90%;
              width: 26rem;
              margin: 20px auto;
              vertical-align: middle;
              display: inline-block;
            }

            .backdrop:before {
              display: inline-block;
              width: 0;
              height: 100%;
              vertical-align: middle;
              content: '';
            }

            .layer {
              position: fixed;
              top: 0;
              left: 0;
              right: 0;
              bottom: 0;
              width: 100%;
              height: 100%;
              opacity: 0.25;
              background-color: black;
              transition: opacity 0.35s cubic-bezier(0.4, 0, 0.2, 1);
              pointer-events: none;
              z-index: 1000;
            }

            .backdrop-wrapper-enter .layer {
              opacity: 0;
            }

            .backdrop-wrapper-enter-active .layer {
              opacity: 0.25;
            }

            .backdrop-wrapper-leave .layer {
              opacity: 0.25;
            }

            .backdrop-wrapper-leave-active .layer {
              opacity: 0;
            }
          </style></div>"
`;
